src/lib/coda_networking/coda_networking.ml

```diff
@@ -423,25 +423,62 @@ module Make (Inputs : Inputs_intf) = struct
     Gossip_net.query_peer t.gossip_net peer
       Rpcs.Transition_catchup.dispatch_multi state_hash
 
-  (* TODO: Make this rpc `Get_ancestry` more intelligent #1324 *)
-  let get_ancestry t input =
-    let peers = random_peers t 8 in
-    find_map' peers ~f:(fun peer ->
-        match%map
-          Gossip_net.query_peer t.gossip_net peer
-            Rpcs.Get_ancestry.dispatch_multi input
-        with
-        | Ok (Some ancestors) -> Ok ancestors
-        | Ok None ->
-            Or_error.errorf
-              !"Peer %{sexp:Peer.t} does not have proof for \
-                %{sexp:Rpcs.Get_ancestry.query}"
-              peer input
-        | Error e ->
-            Or_error.errorf
-              !"Encountered a network connection querying an ancestry from \
-                peer %{sexp:Peer.t}: %{sexp:Error.t}"
-              peer e )
+  let get_ancestry_non_preferred_peers t input =
+    let max_peers = 8 in
+    let rec loop num_peers =
+      if num_peers > max_peers then
+        return
+          (Or_error.errorf
+             !"None of randomly-chosen peers has a proof for \
+               %{sexp:Rpcs.Get_ancestry.query}"
+             input)
+      else
+        let peers = random_peers t num_peers in
```

Comment:
This `loop` can query the same peers. It would not be efficient for this to happen because a peer may not have a external_transition. Can you change this loop so that we do not have to query the same peers all the time?